### DISCUSSION TITLE
Row count in Oracle

### DISCUSSION THREAD

**Discussion by explorercpp** [oracle/python-oracledb#247] [Q&A] [answered]
Good afternoon,

For example if I run simple test and select e.g. 400 records. I'm trying to confirm in the network packet, that there were 400 records received back. 
Is there some value encoded in the beginning or in the end of the response from the database server, that will say 400?

**Comment by explorercpp** [comment]
```
00000000 : 00 00 00 77 06 00 00 00 00 00 03 5e 05 02 80 61 ...w.......^...a
00000010 : 00 01 01 29 01 01 0d 00 00 00 01 02 04 7f ff ff ...)............
00000020 : ff 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 ................
00000030 : 00 00 00 00 00 00 00 00 00 00 00 00 29 53 45 4c ............)SEL
00000040 : 45 43 54 20 2a 20 46 52 4f 4d 20 63 6f 75 6e 74 ECT * FROM count
00000050 : 65 72 20 57 48 45 52 45 20 52 4f 57 4e 55 4d 20 er WHERE ROWNUM
00000060 : 3c 3d 20 34 30 30 01 01 00 00 00 00 00 00 01 01 <= 400..........
00000070 : 00 02 80 00 00 00 00                            .......
```

this is the beginning of the response:
```
00000000 : 00 00 01 2d 06 00 00 00 00 00 10 17 40 1b 65 5d ...-........@.e]
00000010 : f3 8e 36 40 6b 05 37 83 da de 46 99 78 7b 0a 13 ..6@k.7...F.x{..
00000020 : 0f 12 27 01 5c 01 03 5c 02 00 0a 00 01 16 00 00 ..'.\..\........
00000030 : 00 00 00 00 00 00 00 0b 01 0b 0b 45 4d 50 4c 4f ...........EMPLO
00000040 : 59 45 45 5f 49 44 00 00 00 00 01 80 00 00 01 32 YEE_ID.........2
00000050 : 00 00 00 00 02 03 69 01 01 32 02 3f fe 00 0d 01 ......i..2.?....
00000060 : 0d 0d 45 4d 50 4c 4f 59 45 45 5f 4e 41 4d 45 00 ..EMPLOYEE_NAME.
00000070 : 00 01 01 00 01 80 00 00 01 14 00 00 00 00 02 03 ................
00000080 : 69 01 01 14 02 3f fe 01 0d 01 0d 0d 45 4d 50 4c i....?......EMPL
00000090 : 4f 59 45 45 5f 43 41 52 44 00 00 01 02 00 01 07 OYEE_CARD.......
000000a0 : 07 78 7b 0a 13 0f 12 32 01 01 02 1f e8 01 0a 01 .x{....2........
000000b0 : 0a 00 06 22 01 03 00 01 02 00 00 00 07 02 c1 02 ..."............
000000c0 : 04 4e 69 6b 65 12 32 30 31 34 20 30 30 30 30 20 .Nike.2014 0000
000000d0 : 30 30 30 30 20 30 30 39 07 02 c1 02 04 4e 69 6b 0000 009.....Nik
000000e0 : 65 13 35 35 30 30 20 30 30 30 30 20 30 30 30 30 e.5500 0000 0000
000000f0 : 20 30 30 30 34 08 01 06 04 07 5a 4d a7 00 01 08  0004.....ZM....
00000100 : 00 00 00 00 00 00 04 01 01 02 24 92 01 02 00 00 ..........$.....
00000110 : 00 01 08 01 0e 03 00 00 00 00 00 00 00 00 00 00 ................
00000120 : 00 00 05 00 01 01 00 00 00 00 00 01 02          .............
```

and this is the end of the response:
```
00000950 : 30 20 30 30 30 34 04 01 01 02 24 96 02 01 90 02 0 0004....$.....
00000960 : 05 7b 00 00 01 08 00 03 00 00 00 00 00 00 00 00 .{..............
00000970 : 00 00 00 00 09 00 00 00 00 00 00 02 05 7b 02 01 .............{..
00000980 : 90 19 4f 52 41 2d 30 31 34 30 33 3a 20 6e 6f 20 ..ORA-01403: no
00000990 : 64 61 74 61 20 66 6f 75 6e 64 0a                data found.
```

**Comment by anthony-tuininga** [comment, marked as answer]
Yes, the value is found in the error record (found at the end of the response). It is the third field (starting at byte 0x95c and consisting of the bytes 0x02, 0x01, 0x90) and indicates the value is 400. Since that field can only go up to 4 bytes (2 ^ 32) the value is *also* found near the end of the record (starting at byte 0x97e and also consisting of the bytes 0x02, 0x01, 0x90) and indicates the value is 400. This value can go up to 8 bytes (2 ^ 64).

The Python driver only pays attention to the second value (for DML) and ignores it for queries -- since by that point the driver has already processed all of the rows and doesn't need to be told how many are there!

**Reply by explorercpp** [reply]
Thank you, that's helpful. I can see it is changing with number of records.
May I also ask, if you might bring some light to information in the beginning of the request
```
00000000 : 00 00 00 77 06 00 00 00 00 00 03 5e 05 02 80 61 ...w.......^...a
00000010 : 00 01 01 29 01 01 0d 00 00 00 01 02 04 7f ff ff ...)............
00000020 : ff 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 ................
00000030 : 00 00 00 00 00 00 00 00 00 00 00 00 29 53 45 4c ............)SEL
00000040 : 45 43 54 20 2a 20 46 52 4f 4d 20 63 6f 75 6e 74 ECT * FROM count
00000050 : 65 72 20 57 48 45 52 45 20 52 4f 57 4e 55 4d 20 er WHERE ROWNUM
00000060 : 3c 3d 20 34 30 30 01 01 00 00 00 00 00 00 01 01 <= 400..........
00000070 : 00 02 80 00 00 00 00                            .......
```

specifically this part on the second line:
`04 7f ff ff`

and this line at the end:
`01 01 00 00 00 00 00 00 01 01 <= 400..........
00000070 : 00 02 80 00 00 00 0`

**Reply by anthony-tuininga** [reply]
The ``04 7f ff ff ff`` value refers to the maximum long size that will be accepted (2 ^ 31).

The stuff after the SQL statement is a 13 element array of integers containing a variety of information. You can find what some of them mean by looking at the Python code:

https://github.com/oracle/python-oracledb/blob/main/src/oracledb/impl/thin/messages.pyx#L2010-L2034

Are you trying to understand the protocol in order to write a different driver? Or is this just for interest?

**Reply by explorercpp** [reply]
we are trying to understand the protocol structure to create network error handling. potentially in the future we might try to implement some client driver for communication with database. if there is some general specification explaining client to server and server to client communication, it would be helpful to understand some network transmission problems in our environment

**Reply by cjbj** [reply]
There is no separate specification.  If you're writing a driver for another programming language, please let us know.

**Reply by explorercpp** [reply]
currently it is only for error handling. we run different activities with sqlplus, jdbc and python-oracledb and try to compare if we see the network having different load. and sometimes we see some unexpected network activities, that we can't explain without looking into network packets.
